File: python/examples/_shared/render.py
from __future__ import annotations

from collections.abc import Iterable
from datetime import datetime


def format_money(minor_units: int | None, currency: str = "USD") -> str:
    if minor_units is None:
        return "N/A"
    return f"{currency} {minor_units / 100:.2f}"


def format_decimal_money(value: str | None, currency: str = "USD") -> str:
    if value in {None, ""}:
        return "N/A"
    return f"{currency} {float(value):.2f}"


def format_percent(value: float | None) -> str:
    if value is None:
        return "N/A"
    return f"{value:.2f}%"


def parse_iso8601(value: str) -> datetime:
    return datetime.fromisoformat(value.replace("Z", "+00:00"))


def format_timestamp(value: datetime | str | None) -> str:
    if value is None:
        return "N/A"
    parsed = parse_iso8601(value) if isinstance(value, str) else value
    return parsed.strftime("%Y-%m-%d %H:%M UTC")


def render_table(headers: list[str], rows: list[list[str]]) -> str:
    widths = [
        max(len(str(cell)) for cell in [header, *[row[index] for row in rows]])
        for index, header in enumerate(headers)
    ]

    def render_row(values: Iterable[str]) -> str:
        return " | ".join(str(value).ljust(widths[index]) for index, value in enumerate(values))

    separator = "-+-".join("-" * width for width in widths)
    rendered_rows = [render_row(headers), separator]
    rendered_rows.extend(render_row(row) for row in rows)
    return "\n".join(rendered_rows)

File: typescript/examples/_shared/render.ts
function pad(value: number): string {
  return String(value).padStart(2, "0");
}

export function formatMoney(minorUnits: number | null | undefined, currency = "USD"): string {
  if (minorUnits == null) {
    return "N/A";
  }
  return `${currency} ${(minorUnits / 100).toFixed(2)}`;
}

export function formatDecimalMoney(
  value: number | string | null | undefined,
  currency = "USD",
): string {
  if (value == null || value === "") {
    return "N/A";
  }

  const parsed = typeof value === "number" ? value : Number.parseFloat(value);
  if (Number.isNaN(parsed)) {
    return "N/A";
  }
  return `${currency} ${parsed.toFixed(2)}`;
}

export function formatPercent(value: number | null | undefined): string {
  if (value == null) {
    return "N/A";
  }
  return `${value.toFixed(2)}%`;
}

export function parseIso8601(value: string): Date {
  const parsed = new Date(value);
  if (Number.isNaN(parsed.getTime())) {
    throw new Error(`Invalid ISO 8601 timestamp: ${value}`);
  }
  return parsed;
}

export function formatTimestamp(value: Date | string | null | undefined): string {
  if (value == null) {
    return "N/A";
  }

  const parsed = typeof value === "string" ? parseIso8601(value) : value;
  return [
    `${parsed.getUTCFullYear()}-${pad(parsed.getUTCMonth() + 1)}-${pad(parsed.getUTCDate())}`,
    `${pad(parsed.getUTCHours())}:${pad(parsed.getUTCMinutes())} UTC`,
  ].join(" ");
}

export function renderTable(
  headers: string[],
  rows: Array<Array<string | number>>,
): string {
  const stringRows = rows.map((row) => row.map((value) => String(value)));
  const widths = headers.map((header, index) =>
    Math.max(header.length, ...stringRows.map((row) => row[index]?.length ?? 0)),
  );

  const renderRow = (values: Array<string | number>): string =>
    values
      .map((value, index) => String(value).padEnd(widths[index], " "))
      .join(" | ");

  const separator = widths.map((width) => "-".repeat(width)).join("-+-");
  return [renderRow(headers), separator, ...stringRows.map(renderRow)].join("\n");
}
